### PR TITLE
feat: impl remote_engine grpc service

### PR DESCRIPTION
**Pull request by chunshao90**
# Which issue does this PR close?

Closes #
Part of #492 

# Rationale for this change
 Implement remote_engine grpc service based on #505.

# What changes are included in this PR?
* Implement remote_engine grpc service.
* Serialize rows with avro.
* Move `ProjectedSchema` to `common.proto`.

# Are there any user-facing changes?
No.

# How does this change test

### CODE REVIEW

File: common_util/src/avro.rs [hunk 1]
```diff
@@ -29,6 +35,21 @@ pub enum Error {
         source: avro_rs::Error,
         backtrace: Backtrace,
     },
+
+    #[snafu(display("Failed to covert to avro record, err:{}", source))]
```
Comment (ShiKaiWi) [inline common_util/src/avro.rs:39]:
```suggestion
    #[snafu(display("Failed to convert to avro record, err:{}", source))]
```

File: common_util/src/avro.rs [hunk 2]
```diff
@@ -29,6 +35,21 @@ pub enum Error {
...
+    ConvertToAvroRecord {
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display(
+        "Invalid avro record, expect record, value:{:?}.\nBacktrace:\n{}",
```
Comment (ShiKaiWi) [inline common_util/src/avro.rs:45]: The error message is confusing, maybe we have to make it clearer.

File: common_util/src/avro.rs [hunk 3]
```diff
@@ -135,6 +156,25 @@ pub fn record_batch_to_avro(
     Ok(())
 }
 
+pub fn convert_record_batch_to_avro(record_batch: RecordBatch) -> Result<Vec<ByteVec>> {
```
Comment (ShiKaiWi) [inline common_util/src/avro.rs:159]: I guess this is a better name: `encode_record_batch_to_avro_rows`.

File: common_util/src/avro.rs [hunk 4]
```diff
@@ -135,6 +156,25 @@ pub fn record_batch_to_avro(
...
+    let mut rows = Vec::new();
+    let avro_schema = to_avro_schema(RECORD_NAME, record_batch.schema());
+    record_batch_to_avro(&record_batch, &avro_schema, &mut rows)?;
+    Ok(rows)
+}
+
+pub fn convert_avro_rows_to_row_group(schema: Schema, rows: &[Vec<u8>]) -> Result<RowGroup> {
```
Comment (ShiKaiWi) [inline common_util/src/avro.rs:166]:
```suggestion
pub fn decode_avro_rows_to_row_group(schema: Schema, rows: &[Vec<u8>]) -> Result<RowGroup> {
```

Comment (ShiKaiWi) [inline common_util/src/avro.rs:166]:
```suggestion
pub fn decode_avro_rows_to_row_group(schema: Schema, rows: &[Vec<u8>]) -> Result<RowGroup> {
```

File: common_util/src/avro.rs [hunk 5]
```diff
@@ -166,3 +241,19 @@ fn may_union(val: Value, is_nullable: bool) -> Value {
         val
     }
 }
+
+fn parse_one_row(schema: &AvroSchema, mut raw: &[u8], row: &mut Vec<Datum>) -> Result<()> {
```
Comment (ShiKaiWi) [inline common_util/src/avro.rs:245]: The tests for newly added methods seems missing, and I guess they should and need these unit tests.

File: proto/protos/remote_engine.proto [hunk 6]
```diff
@@ -61,14 +52,18 @@ message ReadRequest {
...
+  repeated bytes rows = 3;
 }
 
 message RowGroup {
-  common.TableSchema table_schema = 1;
-  repeated bytes rows = 2;
-  int64 min_timestamp = 3;
-  int64 max_timestamp = 4;
+  // Version of row encoding method
+  uint32 version = 1;
```
Comment (ShiKaiWi) [inline proto/protos/remote_engine.proto:62]: Maybe we should put both the `version` and `rows` at the end.

File: table_engine/src/predicate.rs [hunk 7]
```diff
@@ -69,6 +78,28 @@ impl Predicate {
...
+        let mut exprs = Vec::with_capacity(pb.exprs.len());
+        for pb_expr in pb.exprs {
+            let expr = Expr::from_bytes(&pb_expr)
+                .map_err(|e| Box::new(e) as _)
+                .context(ExprDecode)?;
+            exprs.push(expr);
+        }
+        Ok(Self {
+            exprs,
+            time_range: TimeRange::new_unchecked(
```
Comment (ShiKaiWi) [inline table_engine/src/predicate.rs:95]: Shall we do check here?

File: table_engine/src/remote/mod.rs [hunk 8]
```diff
@@ -7,12 +7,46 @@ pub mod model;
...
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display("Failed to covert row group, err:{}", source))]
+    ConvertRowGroup {
+        source: Box<dyn std::error::Error + Send + Sync>,
+    },
+
+    #[snafu(display(
+        "Failed to covert row group, encode version:{}.\nBacktrace:\n{}",
```
Comment (ShiKaiWi) [inline table_engine/src/remote/mod.rs:44]:
```suggestion
        "Failed to covert row group, encoding version:{}.\nBacktrace:\n{}",
```

File: table_engine/src/table.rs [hunk 9]
```diff
@@ -350,6 +387,37 @@ pub struct ReadRequest {
...
+                    .map_err(|e| Box::new(e) as _)
+                    .context(ConvertPredicate)?,
+            ),
+            order: if pb.order == proto::remote_engine::ReadOrder::Asc as i32 {
+                ReadOrder::Asc
+            } else if pb.order == proto::remote_engine::ReadOrder::Desc as i32 {
+                ReadOrder::Desc
+            } else {
+                ReadOrder::None
+            },
```
Comment (ShiKaiWi) [inline table_engine/src/table.rs:416]: Shall split up these fields' initialization? It looks crowded to squeeze all of them in the `Self` initialization.

File: server/src/grpc/remote_engine_service/mod.rs [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+                    if tx
+                        .send(batch.map_err(|e| Box::new(e) as _).context(ErrWithCause {
+                            code: StatusCode::Internal,
+                            msg: "record batch failed",
+                        }))
+                        .await
+                        .is_err()
+                    {
+                        error!("Failed to send handler result.");
+                        break;
```
Comment (Rachelint) [inline server/src/grpc/remote_engine_service/mod.rs:77]:
```suggestion
        for stream in streams.streams {
            let mut stream = stream.map(|result| {
                result.map_err(|e| Box::new(e) as _).context(ErrWithCause {
                    code: StatusCode::Internal,
                    msg: "record batch failed",
                })
            });

            let tx = tx.clone();
            let _ = self.runtimes.read_runtime.spawn(async move {
                while let Some(batch) = stream.next().await {
                    if let Err(e) = tx.send(batch).await {
                        error!("Failed to send handler result, err:{}.", e);
                        break;
                    }
                }
            });
        }
```

File: server/src/grpc/remote_engine_service/mod.rs [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+            .await
+            .map_err(|e| Box::new(e) as _)
+            .context(ErrWithCause {
+                code: StatusCode::Internal,
+                msg: "fail to join task",
+            })??;
+
+        for mut stream in streams.streams.into_iter() {
+            let tx = tx.clone();
+            let _ = self.runtimes.read_runtime.spawn(async move {
```
Comment (Rachelint) [inline server/src/grpc/remote_engine_service/mod.rs:66]: Maybe we should collect the handler and await it in `read` ?

File: server/src/grpc/remote_engine_service/mod.rs [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+    async fn stream_read_internal(
+        &self,
+        request: Request<ReadRequest>,
+    ) -> Result<ReceiverStream<Result<RecordBatch>>> {
+        let ctx = self.handler_ctx();
+        let (tx, rx) = mpsc::channel(STREAM_QUERY_CHANNEL_LEN);
+        let handle = self.runtimes.read_runtime.spawn(async move {
+            let read_request = request.into_inner();
+            handle_stream_read(ctx, read_request).await
+        });
```
[the remark this answers is not in the document]
Comment (chunshao90) [reply server/src/grpc/remote_engine_service/mod.rs:55]: Need to handle `read` in `read_runtime`.

File: common_util/src/avro.rs (hunk 5 above)
Comment (chunshao90) [reply common_util/src/avro.rs:245]: Create an issue to track this. #516

File: common_util/src/avro.rs [hunk 13]
```diff
@@ -158,6 +198,41 @@ fn column_to_value(array: &ColumnBlock, row_idx: usize, is_nullable: bool) -> Va
     }
 }
 
+/// Convert the avro `Value` into the `Datum`.
+///
+/// Some types defined by avro are not used and the conversion rule is totally
+/// based on the implementation in the server.
+fn value_to_datum(value: Value) -> Result<Datum> {
```
Comment (Rachelint) [inline common_util/src/avro.rs:205]: avro_value_to_datum?

File: common_util/src/avro.rs [hunk 14]
```diff
@@ -85,6 +106,25 @@ pub fn to_avro_schema(name: &str, schema: &RecordSchema) -> avro_rs::Schema {
...
+    record_batch_to_avro(record_batch, &avro_schema, &mut rows)?;
+    Ok(rows)
+}
+
+pub fn avro_rows_to_row_group(schema: Schema, rows: &[Vec<u8>]) -> Result<RowGroup> {
+    let avro_schema = to_avro_schema(RECORD_NAME, &schema.to_record_schema());
+    let mut builder = RowGroupBuilder::with_capacity(schema.clone(), rows.len());
+    for raw_row in rows {
+        let mut row = Vec::with_capacity(schema.num_columns());
+        parse_one_row(&avro_schema, raw_row, &mut row)?;
```
Comment (Rachelint) [inline common_util/src/avro.rs:121]: avro_row_to_row

**Review by Rachelint — APPROVED** [review]
LGTM

**Pull request merged**